### ISSUE TITLE
Proposal: Custom rate limiting rules

### ISSUE THREAD

**Issue by MarvinJWendt** [pocketbase/pocketbase#1232]
Preliminary: I know that rate limiting has been proposed before (in #387), but due to the lack of a description, I would like to submit a justified proposal herewith.

## Description

I propose the option to implement basic rate-limiting to specific API endpoints. Most, if not all apps, would benefit from rate-limiting user and record creation. For this proposal, I want to create an imaginary app that is like Instagram.

### Attack vector

Servers / apps will, sooner or later, be attacked. 

1. A common attack vector is to spam the creation of posts (like pictures on our imaginary Instagram clone). At the current state, PocketBase (in its bare form) would allow the creation of thousands of posts per second. This is obviously nothing that we would want in our app.
2. Another attack vector is dumping the whole database. An attacker could query all posts and increase the "page" in a loop. This would cause intense load on the servers memory and the attacker would have every post locally (which might be millions on big apps).
3. The probably most dangerous attack would be to create thousands of users. Those are hard to delete, as it's not tracked which IP created the user. A simple script is currently capable to create users in a loop. Such an attack could hardly be prevented in PocketBases current form.

### Current solution

The current solution requires the use of a reverse-proxy. This would be feasible, but the programmer would have to create individual rules for each route (which are currently not marked as stable, so the rules might need to be updated frequently). With larger apps this is hardly possible to keep synchronous and would require the usage of lots of wildcards (posts creation might have a different rate-limit than post-view for example). Also, some people do not have a reverse-proxy or are not capable to set one up.

### Proposed solution

I propose that an option is added to each collection, how rate-limiting should be handled. This option could be embedded in the existing API rules. A simple implementation could look like this:

> **Note**
> Text in square brackets could be an input field as dropdown or raw input

```
Allow [user/IP/IP||user] [number] [creations/updates/views/lists] per [number] [second/minute/hour] 
```

> **Note**
> The limiting metric IP||User should count request on the same IP or the same user, to prevent users to use proxy lists.
> When anonymous users have access to the collection endpoint, it should fallback to IP||session_id

As a quick mockup, this could look like:
![image](https://user-images.githubusercontent.com/31022056/206884363-d278cbbd-fed2-4162-b182-500925294e94.png)
![image](https://user-images.githubusercontent.com/31022056/206884774-7579789e-4a8c-4f0f-be59-7a926f5a2cd3.png)

Maybe this could be integrated in the API Rule syntax somehow, but I'll leave that up to you :)

This could also be extended, so that unauthorized users, or users which fit a requirement (like a role), have different rate-limits. 

---

Please let me know what you think, in my opinion this would be a very useful security feature!

**Comment by ganigeorgiev** [comment]
While I agree that this would be a nice to have feature, for now it remains "unplanned" because there are way too many other issues with higher priority that represents a blocker to some use cases (multi-match and `@request.data.*` rel filter enhancements, private/protected files, single and combined field indexes, raw sql or collection views for aggregations, etc.).

I'm not even sure that I'll accept a PR if someone decide to work on this. I want to keep PocketBase as simple as possible.
While for the past month I'm lucky enough to be able to [work full-time on PocketBase](https://github.com/pocketbase/pocketbase/discussions/730), I'm well aware that this will not last forever and I have to think a little more on the eventual maintenance burden that will come with every new feature (not to mention that I'm already slightly behind the original planned schedule with my other open source project - [Presentator #183](https://github.com/presentator/presentator/issues/183)).

Maybe the better approach in the long run will be to prioritize the Admin UI extendability (#898) to allow users to create plugins that eventually may get integrated/embedded in the core.

I'll try to reorganize and update the [roadmap](https://github.com/orgs/pocketbase/projects/2) after the upcoming few releases.

**Comment by ganigeorgiev** [comment]
If you want to use the default echo rate-limiter middleware you can register it to the `e.Router` instance in the `OnBeforeServe` event hook:

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
	e.Use(middleware.RateLimiterWithConfig(yourConfig))

	return nil
})
```

You can specify to which routes to apply by using the `Skipper` field of the middleware config.

We can't register middlewares inside other event hooks because they are triggered more than once. Additionally the echo `Use` and `Pre` middleware registration functions are not concurrent safe and may cause some issues if a middleware is registered after the server start (_I haven't really tested it_).

In any case, as mentioned previously for me this personally is a very low priority.
I understand the eventual benefits of it but most users in normal situations don't really need it and in some cases using an in-memory rate-limiter may actually cause more harm than good since it will increase the load on the server and I don't want to rush implementation.

**Comment by MarvinJWendt** [comment]
I made my own custom middleware now, and I think that should be enough for now. If anyone is interested, here is the implementation: 

### Middleware

```go
// RateLimitCollectionOperations is a middleware that rate limits HTTP requests on collection operations.
// Possible operations:
// - list
// - view
// - create
// - update
// - delete
func RateLimitCollectionOperations(collection string, operations []string, limiter echo.MiddlewareFunc) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			contains := func(s string, list []string) bool {
				for _, v := range list {
					if v == s {
						return true
					}
				}
				return false
			}

			collectionEndpoint := fmt.Sprintf("/api/collections/%s/records", collection)

			if contains("list", operations) && c.Request().URL.Path == collectionEndpoint && c.Request().Method == http.MethodGet {
				return limiter(next)(c)
			} else if contains("view", operations) && strings.HasPrefix(c.Request().URL.Path, collectionEndpoint) && c.Request().Method == http.MethodGet {
				return limiter(next)(c)
			} else if contains("create", operations) && strings.HasPrefix(c.Request().URL.Path, collectionEndpoint) && c.Request().Method == http.MethodPost {
				return limiter(next)(c)
			} else if contains("update", operations) && strings.HasPrefix(c.Request().URL.Path, collectionEndpoint) && c.Request().Method == http.MethodPut {
				return limiter(next)(c)
			} else if contains("delete", operations) && strings.HasPrefix(c.Request().URL.Path, collectionEndpoint) && c.Request().Method == http.MethodDelete {
				return limiter(next)(c)
			}

			return next(c)
		}
	}
}
```

### Usage

```go
func main() {
	app := pocketbase.New()

	limiter := middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(2))
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		// Rate limit list and view operations on the "posts" collection
		e.Router.Use(RateLimitCollectionOperations("posts", []string{"list", "view"}, limiter))
		return nil
	})

	app.Start()
}
```

> **Warnng**
I have not tested every case, but for now it seems to work fine. I'll update the solution when I find a bug.
Also, the code could be refactored, but I wanted to keep it a single function for sharing. Feel free to extract the contains method, etc...

**Comment by greendesertsnow** [comment]
We need a login abuse limiter as well

**Comment by MarvinJWendt** [comment]
> We need a login abuse limiter as well

I think the most common approach for login and registration abuse prevention is implementing a captcha. There are ones that can be configured to only prompt the user, if suspicious activities are detected. Even with basic rate limiting, logins could still be brute forced with proxy lists. 

For basic rate limiting registrations, you can use the middleware I posted earlier (with a `create` action on the `users` collection). With some adaptions, this could also be used for login rate limiting, but like I said, this is usually not enough for those "critical public endpoints".

**Comment by ninakylle02** [comment]
Hi

**Comment by GewoonJaap** [comment]
This would be a great addition to PocketBase! Max requests per op/user per action would be amazing!

**Comment by adam-ah** [comment]
While this feature is needed in production, I'm not sure it belongs to Pocketbase itself.
A reverse proxy, such as NGINX, is better suited for geo / rate limiting.

https://www.nginx.com/blog/rate-limiting-nginx/
https://gist.github.com/ipmb/472da2a9071dd87e24d3

**Comment by MarvinJWendt** [comment]
@adam-ah Many highly scalable applications run behind a reverse proxy, of course. But PocketBase is an all-in-one solution. Therefore I think it should not be necessary to put a reverse-proxy in between. Also, some routes need a different rate-limit. When setting up a reverse-proxy you would have to know each API route and limit it individually, which is quite a time-consuming task. Especially when apps get new features, you would have to evaluate each time if the reverse-proxy config needs to be adjusted.

**Comment by pchasco** [comment]
I think the right question to ask is "Is a rate limiter needed?" If the answer is "yes," then in the spirit of PocketBase's tagline, "Open Source backend for your next SaaS and Mobile app in 1 file," then rate limiting should be included in PocketBase.

Personally, I think rate limiting is an important feature, as all services exposed to the internet of any interest at all will suffer some attack. It is certainly as important as some of the nice-to-have features already slated to be implemented.

**Comment by TheRedSpy15** [comment]
To add to alternatives, you can also apply a rate limit on a dns level. Cloudflare makes it easy to configure a WAF

**Comment by adam-ah** [comment]
You cannot apply _"rate limit on a dns level"_, whatever that means, @TheRedSpy15 
Cloudflare is effectively a reverse proxy, and this approach has been discussed before.

**Comment by TheRedSpy15** [comment]
> You cannot apply _"rate limit on a dns level"_, whatever that means, @TheRedSpy15 Cloudflare is effectively a reverse proxy, and this approach has been discussed before.

![image](https://github.com/pocketbase/pocketbase/assets/32081703/0ffffb73-d97d-4dd4-9f6d-324088661cf2)

DNS level referring to using your domain provider.

**Comment by ganigeorgiev** [comment]
It needs some tests but a builtin rate limiter is implemented as part of the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and will be available with the next v0.23.0 release (_there are no hard ETAs but I expect to finalize at least an RC of the new Go APIs within a month or two at worst_).

The planned rate limiter implementation is part of a new "Security" PocketBase Settings section.

The implementation differs a little from the proposal and it is not tightly coupled to the collections in order to allow managing rate limits for custom user defined routes in a unified manner. This is currently done using "tags".

The general idea is that users will have an option to define several "tagged" rate limit 
 configs (for example "users:create" -> 0.5 r/s) which will be then picked by the `apis.RateLimit(optTags)` middleware. In case the middleware has multiple tags listed we retrieve the rate config only for the first found tagged entry (left to right).

_Note that we don't have to pass a concrete PocketBase `core.App` instance when registering the `apis.RateLimit()` middleware because with the new routing (and hooks) all handlers have access to `e.App` field (where `e` is the event handler argument)._

The builtin endpoints have the `RateLimit` middleware registered by default with a predefined tag set (this will be documented additionally).

For example, the create action for the "users" collection will have something like `apis.RateLimit("users:create", "*:create")` (_it is a little more complicated in reality because the collection name can change during runtime_).

All `apis.RateLimit` registration will have the `"*"` tag available by default as a final fallback in case a global tag config is defined, or in other words:
`apis.RateLimit("users:create", "*:create")` is the same as `apis.RateLimit("users:create", "*:create", "*")`

For now the `RateLimit` middleware internally uses the std [`time/rate`](https://pkg.go.dev/golang.org/x/time/rate) "Token bucket" implementation stored in plain map.

Based on my local tests it performs fine and I wasn't able to cause lock contention once replaced the regular `sync.Mutex` with `sync.RWMutex` (_it is slower for new writes but allows multiple readers which is better suited for our case_).

For simplicity the unique client identifier initially will be only the client IP. For the cases where PocketBase is deployed behind a reverse proxy there is an option in the settings for users to explicit mark `X-Forward-For` or other custom header(s) like `CF-Connecting-IP` as "trusted" (_I haven't explored all cases yet but eventually we may also need an option to specify a list of trusted proxy IPs_).

On every rate limiter setting change we reset all collected rates data.

In case the rate limiter is disabled from the settings the `apis.RateLimit()` middleware does nothing.

**Comment by cddqssc** [comment]
> It needs some tests but a builtin rate limiter is implemented as part of the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and will be available with the next v0.23.0 release (_there are no hard ETAs but I expect to finalize at least an RC of the new Go APIs within a month or two at worst_).它需要一些测试，但内置的速率限制器是作为[正在进行的重构](https://github.com/pocketbase/pocketbase/discussions/4355)的一部分实现的，并将在下一个 v0.23.0 版本中提供（_没有硬性的 ETA，但我希望至少在一个新的 Go API 中最终确定一个 RC）最坏的情况是一两个月_）。
> 
> The planned rate limiter implementation is part of a new "Security" PocketBase Settings section.计划的速率限制器实施是新的“安全”PocketBase 设置部分的一部分。
> 
> The implementation differs a little from the proposal and it is not tightly coupled to the collections in order to allow managing rate limits for custom user defined routes in a unified manner. This is currently done using "tags".该实现与提案略有不同，并且它没有与集合紧密耦合，以便允许以统一的方式管理自定义用户定义的路由的速率限制。目前这是使用“标签”完成的。
> 
> The general idea is that users will have an option to define several "tagged" rate limit总体思路是用户可以选择定义几个“标记”速率限制 configs (for example "users:create" -> 0.5 r/s) which will be then picked by the `apis.RateLimit(optTags)` middleware. In case the middleware has multiple tags listed we retrieve the rate config only for the first found tagged entry (left to right).配置（例如“users:create”-> 0.5 r/s），然后由`apis.RateLimit(optTags)`中间件选择。如果中间件列出了多个标签，我们仅检索第一个找到的标记条目（从左到右）的速率配置。
> 
> _Note that we don't have to pass a concrete PocketBase `core.App` instance when registering the `apis.RateLimit()` middleware because with the new routing (and hooks) all handlers have access to `e.App` field (where `e` is the event handler argument).请注意，在注册`apis.RateLimit()`中间件时，我们不必传递具体的 PocketBase `core.App`实例，因为使用新的路由（和挂钩），所有处理程序都可以访问`e.App`字段（其中`e`是事件处理程序）争论）。_
> 
> The builtin endpoints have the `RateLimit` middleware registered by default with a predefined tag set (this will be documented additionally).内置端点默认使用预定义标签集注册`RateLimit`中间件（这将另外记录）。
> 
> For example, the create action for the "users" collection will have something like `apis.RateLimit("users:create", "*:create")` (_it is a little more complicated in reality because the collection name can change during runtime_).例如，“users”集合的创建操作将具有类似的内容 `apis.RateLimit("users:create", "*:create")` （_实际上有点复杂，因为集合名称可以在运行时更改_）。
> 
> All `apis.RateLimit` registration will have the `"*"` tag available by default as a final fallback in case a global tag config is defined, or in other words:默认情况下，所有`apis.RateLimit`注册都将具有`"*"`标记，作为定义全局标记配置时的最终后备，或者换句话说： `apis.RateLimit("users:create", "*:create")` is the same as `apis.RateLimit("users:create", "*:create", "*")` `apis.RateLimit("users:create", "*:create")` 是一样的 `apis.RateLimit("users:create", "*:create", "*")`
> 
> For now the `RateLimit` middleware internally uses the std [`time/rate`](https://pkg.go.dev/golang.org/x/time/rate) "Token bucket" implementation stored in plain map.目前， `RateLimit`中间件在内部使用存储在普通映射中的标准[`time/rate`](https://pkg.go.dev/golang.org/x/time/rate) “令牌桶”实现。
> 
> Based on my local tests it performs fine and I wasn't able to cause lock contention once replaced the regular `sync.Mutex` with `sync.RWMutex` (_it is slower for new writes but allows multiple readers which is better suited for our case_).根据我的本地测试，它表现良好，并且一旦用`sync.RWMutex`替换常规`sync.Mutex` ，我就无法导致锁争用（_新写入速度较慢，但​​允许多个读取器，这更适合我们的情况_）。
> 
> For simplicity the unique client identifier initially will be only the client IP. For the cases where PocketBase is deployed behind a reverse proxy there is an option in the settings for users to explicit mark `X-Forward-For` or other custom header(s) like `CF-Connecting-IP` as "trusted" (_I haven't explored all cases yet but eventually we may also need an option to specify a list of trusted proxy IPs_).为简单起见，唯一的客户端标识符最初只是客户端 IP。对于 PocketBase 部署在反向代理后面的情况，设置中有一个选项可供用户显式标记`X-Forward-For`或其他自定义标头（如`CF-Connecting-IP`为“可信”（_我没有探索了所有情况，但最终我们可能还需要一个选项来指定受信任代理 IP 列表_）。
> 
> On every rate limiter setting change we reset all collected rates data.每次速率限制器设置更改时，我们都会重置所有收集的速率数据。
> 
> In case the rate limiter is disabled from the settings the `apis.RateLimit()` middleware does nothing.如果从设置中禁用速率限制器，则`apis.RateLimit()`中间件不会执行任何操作。

[v0.23.0-rc13 Prerelease](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.0-rc13)rateLimit in middlewares_rate_limit.go seems to be private and cannot be called directly. Will it be made public in the future?

**Comment by ganigeorgiev** [comment]
@cddqssc The rate limiter implementation has changed since the https://github.com/pocketbase/pocketbase/issues/1232#issuecomment-2272763235 announcement.

The rate limiter now is globally registered and it is controlled from the _Dashboard > Settings > Applications > Rate limiting._
